src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/CompletionService.cs

```diff
@@ -0,0 +1,172 @@
+        /// <summary>
+        /// Created new instance given binding queue
+        /// </summary>
+        public CompletionService(ConnectedBindingQueue bindingQueue)
+        {
+            BindingQueue = bindingQueue;
+        }
+
+        private ISqlParserWrapper sqlParserWrapper;
+
+        /// <summary>
+        /// SQL parser wrapper to create the completion list
+        /// </summary>
+        public ISqlParserWrapper SqlParserWrapper
+        {
+            get
+            {
+                if(this.sqlParserWrapper == null)
+                {
+                    this.sqlParserWrapper = new SqlParserWrapper();
+                }
+                return this.sqlParserWrapper;
+            }
+            set
+            {
+                this.sqlParserWrapper = value;
+            }
+        }
+
+        /// <summary>
+        /// Creates a completion list given connection and document info
+        /// </summary>
+        public AutoCompletionResult CreateCompletions(
+            ConnectionInfo connInfo,
+            ScriptDocumentInfo scriptDocumentInfo,
+            bool useLowerCaseSuggestions)
+        {
+            AutoCompletionResult result = new AutoCompletionResult();
+            // check if the file is connected and the file lock is available
+            if (scriptDocumentInfo.ScriptParseInfo.IsConnected && Monitor.TryEnter(scriptDocumentInfo.ScriptParseInfo.BuildingMetadataLock))
+            {
+                try
+                {
+                    QueueItem queueItem = AddToQueue(connInfo, scriptDocumentInfo.ScriptParseInfo, scriptDocumentInfo, useLowerCaseSuggestions);
+
+                    // wait for the queue item
+                    queueItem.ItemProcessed.WaitOne();
+                    var completionResult = queueItem.GetResultAsT<AutoCompletionResult>();
+                    if (completionResult != null && completionResult.CompletionItems != null)
+                    {
+                        result = completionResult;
+                    }
+                    else if (!ShouldShowCompletionList(scriptDocumentInfo.Token))
+                    {
+                        result.CompleteResult(scriptDocumentInfo.ScriptParseInfo.CurrentSuggestions, AutoCompleteHelper.EmptyCompletionList);
+                    }
+                }
+                finally
+                {
+                    Monitor.Exit(scriptDocumentInfo.ScriptParseInfo.BuildingMetadataLock);
+                }
+            }
+
+            return result;
+        }
+
+        private QueueItem AddToQueue(
+            ConnectionInfo connInfo,
+            ScriptParseInfo scriptParseInfo,
+            ScriptDocumentInfo scriptDocumentInfo,
+            bool useLowerCaseSuggestions)
+        {
+            // queue the completion task with the binding queue    
+            QueueItem queueItem = this.BindingQueue.QueueBindingOperation(
+                key: scriptParseInfo.ConnectionKey,
+                bindingTimeout: 60000,
```

Comment:
>60000 [](start = 32, length = 5)

the timeout is going to be different based on which type of operation is being queued.  this should be a parameter somewhere.